### PR TITLE
Eager Casper

### PR DESCRIPTION
**Pull request by nzpr**
## Overview
Initial implementation of Casper operating on a local view of the message.

### Please make sure that this PR:
- [x] is at most 200 lines of code (excluding tests),
- [x] meets [RChain development coding standards](https://rchain.atlassian.net/wiki/spaces/DOC/pages/28082177/Coding+Standards),
- [x] includes tests for all added features,
- [x] has a reviewer assigned,
- [x] has [all commits signed](https://rchain.atlassian.net/wiki/spaces/DOC/pages/498630673/How+to+sign+commits+to+rchain+rchain).

### [Bors](https://bors.tech/) cheat-sheet:

- `bors r+` runs integration tests and merges the PR (if it's approved),
- `bors try` runs integration tests for the PR,
- `bors delegate+` enables non-maintainer PR authors to run the above.

### CODE REVIEW

File: shared/src/main/scala/coop/rchain/casper/lcasper/FinalityView.scala [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+    //  therefore the highest one should be chosen.
+    val highestCompleteLevelIdx = targetViews.map(_.completeLevelIdx).max
+    val views = targetViews.map {
+      case v @ FinalityView(levels) =>
+        // align all views to start from the same level
+        val diff = highestCompleteLevelIdx - v.completeLevelIdx
+        FinalityView(levels.drop(diff.toInt))
+    }
+    assert(
+      views.map(_.completeLevelIdx).toSet.size == 1,
```
Comment (tgrospic) [inline shared/src/main/scala/coop/rchain/casper/lcasper/FinalityView.scala]: Use `distinct` if you don't need Set instance.

It faster see https://github.com/rchain/rchain/pull/3568#discussion_r802092553.

File: shared/src/main/scala/coop/rchain/casper/lcasper/FinalityView.scala [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+    )
+    // align views bu adding dummy records on top
+    val maxHeight = views.map(_.levelsViews.size).max
+    val aligned = views.map { v =>
+      val toFill  = maxHeight - v.levelsViews.size
+      val padding = (1 to toFill).map(_ => LevelView(-1, Vector.empty[Fringe[M, S]]))
+      v.copy(levelsViews = v.levelsViews ++ padding)
+    }
+
+    // reconcile top down
```
Comment (tgrospic) [inline shared/src/main/scala/coop/rchain/casper/lcasper/FinalityView.scala]: What top down means here? Should we reconcile from common fringe?

File: shared/src/main/scala/coop/rchain/casper/lcasper/LevelView.scala [hunk 3]
```diff
@@ -0,0 +1,78 @@
+package coop.rchain.casper.lcasper
+import coop.rchain.casper.lcasper.Casper.{Fringe, LevelReconciliation}
+
+import scala.annotation.tailrec
+
+// View on DAG level
```
Comment (tgrospic) [inline shared/src/main/scala/coop/rchain/casper/lcasper/LevelView.scala]: What do you mean by _DAG level_?

File: shared/src/main/scala/coop/rchain/casper/lcasper/LevelView.scala [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+      bondsMap: Map[S, Long]
+  ): LevelReconciliation[M, S] = {
+    require(
+      views.map(_.height).distinct.size == 1,
+      "Unexpected error: reconciliation across different levels is detected."
+    )
+
+    final case class PastureState(
+        amoeba: Vector[(Fringe[M, S], Int)],          // indexed legs of amoeba
+        pasture: Vector[Vector[(Fringe[M, S], Int)]], // indexed pieces of food in the pasture
```
Comment (tgrospic) [inline shared/src/main/scala/coop/rchain/casper/lcasper/LevelView.scala]: What is the meaning of index? Vector has index of elements and Int is in tuple, which index you are referring to?

File: shared/src/main/scala/coop/rchain/casper/lcasper/FinalityView.scala (hunk 2 above)
Comment (nzpr) [reply shared/src/main/scala/coop/rchain/casper/lcasper/FinalityView.scala]: I think it is not relevant for this particular piece. Here we just collapse two view.
```scala
final case class LevelView[M, S](height: Long, partitions: Vector[Fringe[M, S]]) 
```
Level view consists of non overlapping partitions (partial fringes), inside which conflicts are already resolved. 
The outcome of reconciliation of two levels can be just placing one `partitions` next to another, or, if some partitions are  overlapping - placing one non overlapping  `partitions` next to another PLUS placing next to this new bigger partitions. 

After this, conflict resolution should be done from bottom up, as you've said.
New rejections done by lower level should be respected when doing resolution in higher levels. 
So `FinalityReconciliation` computed here should be input for conflict resolver, which then should go from the common fringe.

File: shared/src/main/scala/coop/rchain/casper/lcasper/LevelView.scala (hunk 3 above)
Comment (nzpr) [reply shared/src/main/scala/coop/rchain/casper/lcasper/LevelView.scala]: Block height (blockNum)

File: shared/src/main/scala/coop/rchain/casper/lcasper/LevelView.scala (hunk 4 above)
Comment (nzpr) [reply shared/src/main/scala/coop/rchain/casper/lcasper/LevelView.scala]: Input for this algo is list of level views, which are list of non overlapping partitions. 
So when partition Psmall from view V1 is absorbed by Pbig in V2 - we should remove absorbed partition from V1 and replace partition with the aggregate one in V2. To be able to do this we need to be able to remove item with some index from a vector and replace item by its index. So this index is used for this purpose.

File: shared/src/main/scala/coop/rchain/casper/lcasper/FinalityView.scala (hunk 2 above)
Comment (nzpr) [reply shared/src/main/scala/coop/rchain/casper/lcasper/FinalityView.scala]: But I just thought that you are right. If we go bottom up, at some level we can decide that LevelView's cannot overlap anymore, and apply some special case to combine remaining views as non overlapping.

**Comment by nzpr** [comment]
Closed in favour of https://github.com/rchain/rchain/pull/3606

**Pull request closed without merging**